### ISSUE TITLE
torch.compile + aot_eager: backward ignores inner autocast(enabled=False) vs eager

### ISSUE THREAD

**Issue by yanboliang** [pytorch/pytorch#179804]
## Repro
```
import torch
import torch.nn as nn

w = nn.Linear(64, 64, device="cuda")
h = torch.randn(4, 64, device="cuda", dtype=torch.bfloat16, requires_grad=True)

def inner_fn(x):
    with torch.amp.autocast("cuda", enabled=False):
        return x.float() @ w.weight.float().T

def outer_fn(x, compile=False):
    if compile:
        return torch.compile(inner_fn, backend="aot_eager")(x)
    return inner_fn(x)

with torch.amp.autocast("cuda", dtype=torch.bfloat16):
    out_eager = outer_fn(h)
with torch.amp.autocast("cuda", dtype=torch.bfloat16):
    out_compiled = outer_fn(h, compile=True)

out_eager.sum().backward()
grad_eager = w.weight.grad.clone()

w.weight.grad = None
out_compiled.sum().backward()
grad_compiled = w.weight.grad.clone()

assert torch.equal(grad_eager, grad_compiled)  # fails
```

## Expected
Eager and compiled paths should bitwise match: forward runs under outer autocast(bf16) but the matmul is forced to fp32 via inner autocast(enabled=False); backward is outside autocast, so weight gradients should agree (fp32 in both cases for this setup).

## Actual
grad_eager and grad_compiled differ (torch.equal is False).

## Hypothesis (short)
In eager mode, autograd restores the per-op autocast state from forward, so backward matmuls run without the outer bf16 autocast. When compile with `aot_eager` backend, backward is traced under the outer autocast via aot autograd, so bf16 casts can get baked into the compiled backward even when backward() runs outside autocast—so gradients no longer match eager.

### Versions

torch 2.9.1

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @chauhang @penguinwu @bdhirsh @aorenste

**Comment by jansel** [comment]
As a workaround you could do: `torch._functorch.config.patch(backward_pass_autocast="off")`

AI summary of what a full fix for this would look like:

The proper fix is to stop tracing the backward under a single compile-time autocast assumption.

  Right now the problematic logic is in torch/_functorch/_aot_autograd/graph_capture_wrappers.py:419. During forward-time joint tracing, AOTAutograd does
  this:

  - "same_as_forward": keep the current ambient autocast state
  - "off": disable autocast
  - custom list: enable that explicit autocast context

  That means the backward graph is baked once, before backward() ever runs. For your repro, that bakes bf16 casts into the backward even though the real
  backward() happens outside autocast.

  What a real fix should be

  Move from “trace backward at forward-compile time” to “trace backward lazily under the actual backward autocast state.”

  Concretely:

  1. Keep the forward trace as-is.
  2. Change lazy backward from “compile a pre-traced bw_module later” to “trace and compile backward on first backward().”
  3. Key the cached backward variant by the actual autocast TLS, e.g.:
      - CUDA enabled
      - CPU enabled
      - CUDA dtype
      - CPU dtype
      - cache enabled
  4. If the same compiled forward is later used under a different backward autocast state, retrace a second backward variant and cache it too.

  That fixes this issue without changing semantics for existing "on", "off", or custom cases.

  Why this is the right design

  The current model fundamentally lacks information: if you compile only a forward slice, the compiler cannot know at compile time whether backward() will
  run:

  - inside the same autocast region,
  - outside autocast,
  - or inside some different autocast context.

  So any single baked backward policy is inherently lossy. The correct solution is to bind backward tracing to the actual runtime backward context.

  What not to do

  - Don’t just add a runtime error guard. That avoids silent wrong results but doesn’t fix behavior.
  - Don’t change "same_as_forward" globally to "off". That breaks existing valid use cases and tests.
  - Don’t try to special-case only autocast(enabled=False). The underlying issue is broader: backward context is dynamic.

  Where the implementation would land

  - torch/_functorch/_aot_autograd/graph_capture_wrappers.py:419: remove the compile-time autocast assumption from joint backward tracing.
  - torch/_functorch/_aot_autograd/runtime_wrappers.py:2643: extend lazy backward so it can retrace, not just compile, keyed on actual backward autocast
    state.
  - The lazy backward info object would need enough context to regenerate the backward graph, not only compile an already-created bw_module.

  Short version

  The proper fix is: lazy retrace backward at first backward() under the real autocast state, and cache per-state backward graphs. That’s the clean fix for
  this bug class.

**Comment by shino16** [comment]
Hi, let me work on this issue as this looks particularly interesting to me.

**Comment by shino16** [comment]
This is a duplicate of #153044 and the fix is proposed in https://github.com/pytorch/pytorch/pull/177286.

**Comment by XAheli** [comment]
Ig we can close this? @shino16 @kshitij12345
